File: Homework/HW10/JAVA/PrintDirectoryStructure/src/edu/citadel/cs/PrintDirectoryStructure.java
package edu.citadel.cs;


import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;


/**
 * Utility class that prints the directory structure to standard output showing the composition of
 * nested files and subdirectories.
 */
public class PrintDirectoryStructure {

  private static final String DIRECTORY_SYMBOL = "+";
  private static final String FILE_SYMBOL = "-";

  public static void printTree(Path file, PrintDirectoryStructureFileVisitor visitor)
      throws IOException {
    Files.walkFileTree(file, visitor);
  }

  protected static void printDirectory(Path dir, int nestingLevel) {
    System.out.println(toString(dir, nestingLevel, DIRECTORY_SYMBOL));
  }


  protected static void printFile(Path file, int nestingLevel) {
    System.out.println(toString(file, nestingLevel, FILE_SYMBOL));
  }

  private static String toString(Path file, int nestingLevel, String symbol) {
    String indent = getIndentString(nestingLevel);
    return String.format("%s%s %s", indent, symbol, file.getFileName());
  }

  private static String getIndentString(int nestingLevel) {
    StringBuilder s = new StringBuilder();

    for (int i = 0; i < nestingLevel; i++) {
      s.append("   ");
    }

    return s.toString();
  }


  static void printUsage() {
    System.out.println("Usage: java edu.citadel.util.PrintDirectoryStructure(<path>)");
    System.out.println("       where <path> is the path of a file or directory");
    System.out.println();
  }
}

File: Homework/HW3/Dart/PrintDirectoryStructure/lib/edu.citadel.util/print_directory_structure.dart
import 'dart:io';
import 'package:path/path.dart' as path;

/// Utility class that prints the directory structure to standard output
/// showing the composition of nested files and subdirectories.
class PrintDirectoryStructure {
  static var _nestingLevel = 0;

  //In Dart, FileSystemEntity is the super class to File, Directory, and Link
  static void printTree(FileSystemEntity entity) {
    if (FileSystemEntity.isDirectorySync(entity.path)) {
      _printDirectory(entity, _nestingLevel);
      _nestingLevel++;

      (entity as Directory).listSync().forEach((entity) => printTree(entity));

      _nestingLevel--;
    } else if (FileSystemEntity.isFileSync(entity.path)) {
      _nestingLevel++;

      _printFile(entity, _nestingLevel);

      _nestingLevel--;
    }
  }

  static void _printDirectory(FileSystemEntity entity, int nestingLevel) {
    stdout.writeln(_getEntityName(entity, nestingLevel, '+'));
  }

  static void _printFile(FileSystemEntity entity, int nestingLevel) {
    stdout.writeln(_getEntityName(entity, nestingLevel, '-'));
  }

  static String _getEntityName(
      FileSystemEntity entity, int nestingLevel, String symbol) {
    var indent = _getIndentString(nestingLevel);
    return '${indent}${symbol} ${path.basename(entity.path)}';
  }

  static String _getIndentString(int nestingLevel) {
    var s = StringBuffer();

    for (var i = 0; i < nestingLevel; i++) {
      s.write('   ');
    }

    return s.toString();
  }

  static void printUsage() {
    stdout.writeln(
        'Usage: dart edu.citadel.util.PrintDirectoryStructure(<path>)');
    stdout.writeln('       where <path> is the path of a file or directory');
    stdout.writeln();
  }
}
